### PR TITLE
Add Swift version

### PR DESCRIPTION
**Pull request by baarde**
## Description

A simple Swift implementation
## Usage

Build with `xcrun --sdk macosx swiftc -O -o csa_swift csa.swift`

Run the test with `ruby test.rb ./csa_swift`
## Limitations
- Loading the data is slow because each line is split into an array of strings before parsing the values
- Not as fast as the reference implementation
- Requires OS X (and Xcode)

### CODE REVIEW

**Comment by Tristramg** [comment]
I leave this to @n-b who will be able to have an opinion on this :)

**Comment by n-b** [comment]
@baarde haha, I can’t say I’m surprised to see you here :)

File: csa.swift [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    let from: Station, to: Station, departure: time_t, arrival: time_t
+    
+    init?(input: InputType) throws {
+        guard let values = try input.read(4) else { return nil }
+        from = values[0]
+        to = values[1]
+        departure = values[2]
+        arrival = values[3]
+        guard from >= 0 && to >= 0 else { throw Error("Station number cannot be negative.") }
+        guard arrival >= departure else { throw Error("Time travel is not supported yet.") }
```
Comment (raphaelmor) [inline csa.swift:26]: :heart:

File: csa.swift [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+do {
+    let input = TextualInput({ readLine() })
+    let router = try Router(input: input)
+    while let request = try Request(input: input) {
+        let route = router.solve(request)
+        printRoute(route)
+    }
+} catch let error as Error {
+    fputs("\(error.message)\n", stdout)
```
Comment (n-b) [inline csa.swift]: It’s not really part of the spec, but `stderr` would probably be more appropriate here.

Comment (baarde) [inline csa.swift]: Silly me, confusing `stderr` and `stdout`! At least I didn't use `stdin`… 😁

File: csa.swift [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        var incoming: [Connection?] = Array(count: count, repeatedValue: nil)
+        for connection in connections {
+            if connection.departure < departure || connection.to == from {
+                continue
+            }
+            if let final = incoming[to] where connection.departure > final.arrival {
+                break
+            }
+            if connection.from != from {
+                guard let previous = incoming[connection.from] where connection.departure > previous.arrival else {
```
Comment (n-b) [inline csa.swift]: 😱 “if not guard let where else”

(this `guard let where else` could be replaced by a regular `if let where`, right?)

File: csa.swift [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        self.count = connections.reduce(0) { (count, connection) in
+            max(count, connection.from + 1, connection.to + 1)
+        }
+    }
+    
+    func solve(request: Request) -> [Connection] {
+        let from = request.from, to = request.to, departure = request.departure
+        guard count > max(from, to) else { return [] }
+        var incoming: [Connection?] = Array(count: count, repeatedValue: nil)
+        for connection in connections {
```
Comment (n-b) [inline csa.swift:61]: You’ve reversed all the conditions from the reference implementations 😁.

I kinda like the pattern `if let <thing> where <condition on thing>`, it’s somewhat self-explaining.

On the other hand the loop itself has C-style imperative logic (`continue`,`break`) and I feel those Swift-y patterns don’t really help in that case. What’s your opinion on this?

**Comment by n-b** [comment]
NOM NOM NOM 😁 capture d ecran 2015-10-28 a 11 59 12

Not really surprising, and the real memory usage is probably less than what’s shown here. But still. 😁

@Tristramg should we track the memory usage in bench.rb as well?

**Comment by Tristramg** [comment]
too difficult to measure :)
But indeed in production with 30 times more data it might hurt a bit :)

File: csa.swift (hunk 3 above)
Comment (baarde) [inline csa.swift]: You mean something like that?

```
if let previous = incoming[connection.from] where connection.departure > previous.arrival {
} else {
    continue
}
```

Comment (n-b) [inline csa.swift]: Yes, but also: why the empty `if` block? why not just reverse the condition and get rid of the `else`?

Comment (baarde) [inline csa.swift]: What would be the negation of `let previous = incoming[connection.from] where connection.departure > previous.arrival`?

You cannot negate a pattern. I can write `incoming[connection.from] == nil || connection.departure <= incoming[connection.from]?.arrival` but I don't like it much.

**Pull request merged**